### PR TITLE
Notifications implementation

### PR DESCRIPTION
**Pull request by ice-myles**
announcement, PingUser, NotifyUser implementation

[task](https://www.notion.so/leftclick/Push-Email-Notifications-295968a3eee34f27832c29ce62f53b94)

### CODE REVIEW

File: notifications/DDL.lua [hunk 1]
```diff
@@ -3,7 +3,8 @@ box.execute([[CREATE TABLE IF NOT EXISTS users  (
                     created_at UNSIGNED NOT NULL,
                     updated_at UNSIGNED NOT NULL,
                     last_mining_started_at UNSIGNED,
-                    last_ping_at UNSIGNED,
+                    -- last_mining_ended_at UNSIGNED,
```
Comment (ice-myles) [inline notifications/DDL.lua:6]: Do we need this field for user in the future?

File: notifications/DDL.lua [hunk 2]
```diff
@@ -20,35 +22,35 @@ box.execute([[CREATE TABLE IF NOT EXISTS notification_types  (
...
-                                  ('T18', 'all', 'Level change: `Your level is now 19.`'),
-                                  ('T19', 'inapp,email', 'Role change: `Your role is now ambassador.`'),
-                                  ('T20', 'all', 'Badge unlock: `You have unlocked "ice breaker" badge`'),
-                                  ('T21', 'inapp,email', 'Task completed: `Congratulations, your first mining session has been started.`'),
-                                  ('T22', 'inapp,push', 'Adoption changed: `Base mining rate is now 15/h ice`'),
-                                  ('T23', 'inapp,push', 'Contact joined ice: `Your friend @jdoe joined ice.`'),
-                                  ('T24', 'all', 'TBD: `New login method merged/linked`'),
-                                  ('T25', 'all', 'TBD: `Start pre-staking`')]])
+box.execute([[INSERT INTO notification_types (name, notification_channels, domain_name, description)
+                           VALUES ('T1', '', 'DomainTODO', 'If the t0 referral pinged the user.'),
```
Comment (ice-myles) [inline notifications/DDL.lua:29]: Can we store domain names in the table? At least it is nice option to get them for the notification.

File: cmd/husky-pack/notifications.go [hunk 3]
```diff
@@ -38,38 +39,110 @@ func (s *service) setupNotificationsRoutes(router *server.Router) {
...
+	}
+	if req.Data.Notification.Push != nil {
+		switch {
+		case len(req.Data.Notification.Push.Description) == 0:
+			return server.BadRequest(errors.Errorf("missing push description"), invalidPropertiesErrorCode)
+		case len(req.Data.Notification.Push.ImageURL) == 0:
+			return server.BadRequest(errors.Errorf("missing push image url"), invalidPropertiesErrorCode)
+		case len(req.Data.Notification.Push.Title) == 0:
+			return server.BadRequest(errors.Errorf("missing push title"), invalidPropertiesErrorCode)
+		case len(req.Data.Notification.Push.Token) == 0: // TODO: do we really need this token, if we take from devices then?
```
Comment (ice-myles) [inline cmd/husky-pack/notifications.go:138]: Do we really need this token, if we take push notification tokens from the devices then?

File: notifications/notify_user.go [hunk 4]
```diff
@@ -4,22 +4,122 @@ package notifications
...
+		},
+		Analytics: &AnalyticsNotification{
+			Action: &tracking.Action{
+				Name: "TODO",
+			},
+		},
+		Type: "T1",
+		// TODO: need to be added smth like this. At least santa needs in the requestingUserId info from ping message.
+		// to have information about who make ping to set achievement.
+		// HiddenData: &JSON{"requestingUserId": requestingUserID},
```
Comment (ice-myles) [inline notifications/notify_user.go:72]: Santa needs in the requestingUserId info from ping message to have the information about who make ping to set achievement. Can we add some field like this to send additional data?

File: notifications/notify_user.go [hunk 5]
```diff
@@ -4,22 +4,122 @@ package notifications
...
+	}
 
-	// requestingUserID := requestingUserID(ctx)
-	// requestingUserID has to be equal to users.referred_by where user_id = userID
+	// TODO: should be LastPingCooldownEndedAt be updated here to know the diff time when user has already been pinged or is this the problem of eskimo/another service?
```
Comment (ice-myles) [inline notifications/notify_user.go:75]: Should be `LastPingCooldownEndedAt` be updated here to know the diff time when user has already been pinged or is this the problem of eskimo/another service?

File: news/news_source.go [hunk 6]
```diff
@@ -68,3 +88,22 @@ func (r *repository) sendTaggedNewsSnapshotMessage(ctx context.Context, ss *Tagg
 
 	return errors.Wrapf(<-responder, "failed to send news snapshot message to broker")
 }
+
+func (r *repository) sendNewsAnnouncement(ctx context.Context, notif *notifications.Notification) error {
+	valueBytes, err := json.MarshalContext(ctx, notif)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal %#v", notif)
+	}
+	key := uuid.NewString()
```
Comment (ice-ares) [inline news/news_source.go:97]: use fmt.("newsId:%v~~~language:%v", ss.ID, ss.Language) as key. And u should add an ID in that notifications.Notification struct, and set that one as that as well

File: news/news_source.go [hunk 7]
```diff
@@ -68,3 +88,22 @@ func (r *repository) sendTaggedNewsSnapshotMessage(ctx context.Context, ss *Tagg
...
+	msg := &messagebroker.Message{
+		Headers: map[string]string{"producer": "husky"},
+		Key:     key,
+		Topic:   r.cfg.MessageBroker.Topics[2].Name, // | announcements.
```
Comment (ice-ares) [inline news/news_source.go:101]: remove the obvious comment

File: news/news_source.go [hunk 8]
```diff
@@ -36,13 +37,32 @@ func (s *newsSnapshotSource) notifyUsers(ctx context.Context, snapshot *TaggedNe
 	if ctx.Err() != nil {
 		return errors.Wrap(ctx.Err(), "unexpected deadline while processing message")
 	}
-	//nolint:godox // .
-	// TODO send message to notification channel to start processing.
-	if s.db != nil {
-		log.Info(fmt.Sprintf("notifUsers for %#v", snapshot))
+	notif := notifications.Notification{ // TODO: check what should be send in proper way.
```
Comment (ice-ares) [inline news/news_source.go:40]: this whole struct has to allow you to send notifications based on specific channels. So you should have An array of arrays, I think, to support parallel and also sequential (fallback) logic

File: notifications/DDL.lua [hunk 9]
```diff
@@ -20,35 +22,35 @@ box.execute([[CREATE TABLE IF NOT EXISTS notification_types  (
                                                                   POSITION('inapp',lower(notification_channels)) > 0 or
                                                                   POSITION('email',lower(notification_channels)) > 0 or
                                                                   POSITION('sms',lower(notification_channels)) > 0 or
-                                                                  POSITION('all',lower(notification_channels)) > 0) and
```
Comment (ice-ares) [inline notifications/DDL.lua:23]: remove the db checks for notification_types

File: notifications/DDL.lua [hunk 10]
```diff
@@ -87,22 +90,23 @@ box.execute([[CREATE TABLE IF NOT EXISTS sent_announcements_history  (
...
                     push_notification_token     STRING,
                     primary key(user_id, device_unique_id)) WITH ENGINE = 'vinyl';]])
-box.execute([[CREATE TABLE IF NOT EXISTS latest_enabled_device_settings (
-                                  disabled_inapp_notification_domains  STRING NOT NULL,
-                                  disabled_email_notification_domains  STRING NOT NULL,
-                                  disabled_sms_notification_domains    STRING NOT NULL,
-                                  language                             STRING NOT NULL DEFAULT 'en',
-                                  user_id                              STRING primary key) WITH ENGINE = 'vinyl';]])
+box.execute([[CREATE TABLE IF NOT EXISTS enabled_named_settings (
+                                  disabled_inapp_notification_domains        STRING NOT NULL,
```
Comment (ice-ares) [inline notifications/DDL.lua:104]: u don't need this

File: notifications/DDL.lua [hunk 11]
```diff
@@ -87,22 +90,23 @@ box.execute([[CREATE TABLE IF NOT EXISTS sent_announcements_history  (
...
+                                  disabled_push_notification_domains         STRING NOT NULL,
+                                  disabled_email_notification_domains        STRING NOT NULL,
+                                  disabled_sms_notification_domains          STRING NOT NULL,
+                                  disabled_analytics_notification_domains    STRING NOT NULL,
```
Comment (ice-ares) [inline notifications/DDL.lua:108]: u don't need this either

File: notifications/DDL.lua [hunk 12]
```diff
@@ -87,22 +90,23 @@ box.execute([[CREATE TABLE IF NOT EXISTS sent_announcements_history  (
...
                     device_unique_id            STRING NOT NULL,
                     push_notification_token     STRING,
                     primary key(user_id, device_unique_id)) WITH ENGINE = 'vinyl';]])
-box.execute([[CREATE TABLE IF NOT EXISTS latest_enabled_device_settings (
-                                  disabled_inapp_notification_domains  STRING NOT NULL,
-                                  disabled_email_notification_domains  STRING NOT NULL,
-                                  disabled_sms_notification_domains    STRING NOT NULL,
-                                  language                             STRING NOT NULL DEFAULT 'en',
-                                  user_id                              STRING primary key) WITH ENGINE = 'vinyl';]])
+box.execute([[CREATE TABLE IF NOT EXISTS enabled_named_settings (
```
Comment (ice-ares) [inline notifications/DDL.lua:103]: move this in user table

File: notifications/DDL.lua (hunk 2 above)
Comment (ice-ares) [reply notifications/DDL.lua:29]: sure

File: notifications/device_metadata_sources.go [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+func (r *repository) modifyDeviceMetadata(ctx context.Context, ds *deviceMetadata) error {
+	if ctx.Err() != nil {
+		return errors.Wrap(ctx.Err(), "context failed")
+	}
+	var resp []*deviceMetadata
+	if err := r.db.UpdateTyped("DEVICE_METADATA", "pk_unnamed_DEVICE_METADATA_1", ds.DeviceID, buildUpdateOps(ds), &resp); err != nil {
+		tErr := new(tarantool.Error)
+		if ok := errors.As(err, tErr); ok && tErr.Code == tarantool.ER_TUPLE_NOT_FOUND { //nolint:nosnakecase // External library.
```
Comment (ice-ares) [inline notifications/device_metadata_sources.go:130]: there's a function in storage for that

File: application.yaml [hunk 14]
```diff
@@ -93,35 +93,40 @@ notifications: &notifications
...
         replicationFactor: 1
         retention: 1000h
-      - name: user-device-metadata-events
+      - name: analytics-announcement-batches
         partitions: 10
         replicationFactor: 1
         retention: 1000h
-      - name: user-device-settings-events
+      ### The next topics are not owned by this service, but are needed to be created for the local/test environment.
+      - name: users-events
```
Comment (ice-ares) [inline application.yaml:105]: use the same name as in the service they come from

File: cmd/husky-pack/notifications.go [hunk 15]
```diff
@@ -38,38 +39,110 @@ func (s *service) setupNotificationsRoutes(router *server.Router) {
...
+func (s *service) SendNotification( //nolint:gocritic,nolint:funlen // .
 	ctx context.Context,
 	req *server.Request[SendNotificationRequestBody, any],
 ) (*server.Response[any], *server.Response[server.ErrorResponse]) {
 	if PingCustomNotificationType == req.Data.CustomType {
 		if err := s.notificationsProcessor.PingUser(ctx, req.Data.UserID); err != nil {
 			err = errors.Wrapf(err, "failed to ping user for %#v", req.Data)
 			switch {
-			// TODO error handle.
+			case errors.Is(err, notifications.ErrNotFound):
```
Comment (ice-ares) [inline cmd/husky-pack/notifications.go:50]: should be relationship not found

File: cmd/husky-pack/notifications.go [hunk 16]
```diff
@@ -38,38 +39,110 @@ func (s *service) setupNotificationsRoutes(router *server.Router) {
...
-	if req.Data.Notification != nil {
-		if err := s.notificationsProcessor.NotifyUser(ctx, req.Data.UserID, req.Data.Notification); err != nil {
-			err = errors.Wrapf(err, "failed to notify user for %#v", req.Data)
-			switch {
-			// TODO error handle.
-			default:
-				return nil, server.Unexpected(err)
-			}
+
+	if err := verifyAdminRole(req.AuthenticatedUser.Role); err != nil {
```
Comment (ice-ares) [inline cmd/husky-pack/notifications.go:64]: follow the same pattern as with news, same for validations

File: cmd/husky-pack/notifications.go [hunk 17]
```diff
@@ -38,38 +39,110 @@ func (s *service) setupNotificationsRoutes(router *server.Router) {
...
+	}
+	if err := validateSendNotification(req); err != nil {
+		return nil, err
+	}
+	if err := s.notificationsProcessor.NotifyUser(ctx, req.Data.UserID, req.Data.Notification); err != nil {
+		err = errors.Wrapf(err, "failed to notify user for %#v", req.Data)
+		switch {
+		case errors.Is(err, notifications.ErrNotFound):
+			return nil, server.NotFound(err, userNotFoundErrorCode)
+		case errors.Is(err, notifications.ErrDuplicate):
```
Comment (ice-ares) [inline cmd/husky-pack/notifications.go:75]: how is this possible?

File: cmd/husky-pack/notifications.go [hunk 18]
```diff
@@ -38,38 +39,110 @@ func (s *service) setupNotificationsRoutes(router *server.Router) {
...
+	if err := verifyAdminRole(req.AuthenticatedUser.Role); err != nil {
+		return nil, err
+	}
+	if err := validateSendNotification(req); err != nil {
+		return nil, err
+	}
+	if err := s.notificationsProcessor.NotifyUser(ctx, req.Data.UserID, req.Data.Notification); err != nil {
+		err = errors.Wrapf(err, "failed to notify user for %#v", req.Data)
+		switch {
+		case errors.Is(err, notifications.ErrNotFound):
```
Comment (ice-ares) [inline cmd/husky-pack/notifications.go:73]: relationship not found. not found is for the main entity, user is a relationship

File: notifications/DDL.lua [hunk 19]
```diff
@@ -20,35 +22,35 @@ box.execute([[CREATE TABLE IF NOT EXISTS notification_types  (
...
-                                  ('T19', 'inapp,email', 'Role change: `Your role is now ambassador.`'),
-                                  ('T20', 'all', 'Badge unlock: `You have unlocked "ice breaker" badge`'),
-                                  ('T21', 'inapp,email', 'Task completed: `Congratulations, your first mining session has been started.`'),
-                                  ('T22', 'inapp,push', 'Adoption changed: `Base mining rate is now 15/h ice`'),
-                                  ('T23', 'inapp,push', 'Contact joined ice: `Your friend @jdoe joined ice.`'),
-                                  ('T24', 'all', 'TBD: `New login method merged/linked`'),
-                                  ('T25', 'all', 'TBD: `Start pre-staking`')]])
+box.execute([[INSERT INTO notification_types (name, notification_channels, domain_name, description)
+                           VALUES ('T1', '', 'DomainTODO', 'If the t0 referral pinged the user.'),
+                                  ('T2', '', 'DomainTODO', 'If mining has been finished for the previous 24 hours: `Start Mining / Don’t forget to start your miner for today.`'),
```
Comment (ice-ares) [inline notifications/DDL.lua:30]: delete T2-T15, T24-T25

Add 'Daily bonus'

Comment (ice-ares) [reply notifications/DDL.lua:30]: check the correct domains from there as well

**Pull request closed without merging**